File: MergeSort.java
import java.io.*;
import java.util.*;

public class MergeSort {
    static class Data {
        int number;
        String str;
        
        Data(int number, String str) {
            this.number = number;
            this.str = str;
        }
    }
    
    public static void main(String[] args) {
        if (args.length < 1 || args.length > 2) {
            System.err.println("Usage: java MergeSort <input_file> [output_file]");
            System.err.println("If output_file is not provided, the output will be written to");
            System.err.println("a file named <input_base>_sorted.csv in the same directory.");
            System.exit(1);
        }
        
        String inputFile = args[0];
        String outputFile;
        
        if (args.length == 2) {
            outputFile = args[1];
        } else {
           
            File inFile = new File(inputFile);
            String baseName = inFile.getName();
            String nameWithoutExtension;
            
            int dotIndex = baseName.lastIndexOf('.');
            if (dotIndex > 0) {  // remove extension if there no first character after dot
                nameWithoutExtension = baseName.substring(0, dotIndex);
            } else {
                nameWithoutExtension = baseName;
            }
            
            outputFile = new File(inFile.getParent(), nameWithoutExtension + "_sorted.csv").toString();
        }
        
        List<Data> data = new ArrayList<>();
        
        // read the input
        long readStart = System.nanoTime();
        try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = br.readLine()) != null) {
                int commaIndex = line.indexOf(',');
                if (commaIndex != -1) {
                    try {
                        int num = Integer.parseInt(line.substring(0, commaIndex));
                        String str = line.substring(commaIndex + 1);
                        data.add(new Data(num, str));
                    } catch (NumberFormatException e) {
                        System.err.println("Skipping invalid number in line: " + line);
                    }
                } else {
                    System.err.println("Skipping malformed line (no comma): " + line);
                }
            }
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            return;
        }
        long readEnd = System.nanoTime();
        
        // array converter for sorting
        Data[] arr = data.toArray(new Data[0]);
        
        // Sort and time
        long sortStart = System.nanoTime();
        mergeSort(arr, 0, arr.length - 1);
        long sortEnd = System.nanoTime();
        
        // output display
        long writeStart = System.nanoTime();
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(outputFile))) {
            for (Data item : arr) {
                bw.write(item.number + "," + item.str);
                bw.newLine();
            }
        } catch (IOException e) {
            System.err.println("Error writing file: " + e.getMessage());
        }
        long writeEnd = System.nanoTime();
        
        // Calculate running time
        double readTime = (readEnd - readStart) / 1e9;  // Convert to seconds
        double sortTime = (sortEnd - sortStart) / 1e9;
        double writeTime = (writeEnd - writeStart) / 1e9;
        double totalTime = readTime + sortTime + writeTime;
        
        System.out.println("Java Merge Sort Times (seconds):");
        System.out.println("Input file: " + inputFile);
        System.out.println("Output file: " + outputFile);
        System.out.printf("Records processed: %,d%n", arr.length);
        System.out.printf("Read time: %.6f s%n", readTime);
        System.out.printf("Sort time: %.6f s%n", sortTime);
        System.out.printf("Write time: %.6f s%n", writeTime);
        System.out.printf("Total time: %.6f s%n", totalTime);
    }
    
    private static void mergeSort(Data[] arr, int l, int r) {
        if (l < r) {
            int m = l + (r - l) / 2;
            mergeSort(arr, l, m);
            mergeSort(arr, m + 1, r);
            merge(arr, l, m, r);
        }
    }
    
    private static void merge(Data[] arr, int l, int m, int r) {
        int n1 = m - l + 1;
        int n2 = r - m;
        Data[] L = new Data[n1];
        Data[] R = new Data[n2];
        
        System.arraycopy(arr, l, L, 0, n1);
        System.arraycopy(arr, m + 1, R, 0, n2);
        
        int i = 0, j = 0, k = l;
        while (i < n1 && j < n2) {
            if (L[i].number <= R[j].number) {
                arr[k++] = L[i++];
            } else {
                arr[k++] = R[j++];
            }
        }
        while (i < n1) arr[k++] = L[i++];
        while (j < n2) arr[k++] = R[j++];
    }
}

File: merge_sort.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <chrono>
#include <iomanip>

using namespace std;
using namespace std::chrono;

struct Data {
    int number;
    string str;
};

void merge(vector<Data>& arr, vector<Data>& temp, int l, int mid, int r) {
    // Left subarray
    int i = l;
    // Right subarray
    int j = mid + 1;
    int k = l;

    while (i <= mid && j <= r) {
        if (arr[i].number <= arr[j].number) {
            temp[k++] = std::move(arr[i++]);
        } else {
            temp[k++] = std::move(arr[j++]);
        }
    }

    while (i <= mid)
        temp[k++] = std::move(arr[i++]);
    while (j <= r)
        temp[k++] = std::move(arr[j++]);

    for (int x = l; x <= r; ++x) {
        arr[x] = std::move(temp[x]);
    }
}

void mergeSort(vector<Data>& arr, vector<Data>& temp, int l, int r) {
    if (l < r) {
        int mid = l + (r - l) / 2;
        mergeSort(arr, temp, l, mid);
        mergeSort(arr, temp, mid + 1, r);
        merge(arr, temp, l, mid, r);
    }
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        cerr << "Usage: " << argv[0] << " <input_file> <output_file>\n"; // can alter input/output when to compile the file
        return 1;
    }
    const string inputFile = argv[1];
    const string outputFile = argv[2];
    vector<Data> data;
    
    // Read input
    auto readStart = high_resolution_clock::now();
    ifstream inFile(inputFile);
    if (!inFile) {
        cerr << "Error opening input file: " << inputFile << endl;
        return 1;
    }
    
    string line;
    while (getline(inFile, line)) {
        size_t pos = line.find(',');
        if (pos != string::npos) {
            try {
                int num = stoi(line.substr(0, pos));
                string str = line.substr(pos + 1);
                data.push_back({num, std::move(str)});
            } catch (...) {
                cerr << "Error parsing line: " << line << endl;
            }
        }
    }
    inFile.close();
    auto readEnd = high_resolution_clock::now();
    
    // Sort data
    auto sortStart = high_resolution_clock::now();
    vector<Data> temp(data.size());
    mergeSort(data, temp, 0, data.size() - 1);
    auto sortEnd = high_resolution_clock::now();
    
    // Write output
    auto writeStart = high_resolution_clock::now();
    ofstream outFile(outputFile);
    if (!outFile) {
        cerr << "Error creating output file: " << outputFile << endl;
        return 1;
    }
    
    for (const auto& item : data) {
        outFile << item.number << "," << item.str << "\n";
    }
    outFile.close();
    auto writeEnd = high_resolution_clock::now();
    
    // Calculate durations
    auto readDuration = duration_cast<microseconds>(readEnd - readStart).count() / 1e6;
    auto sortDuration = duration_cast<microseconds>(sortEnd - sortStart).count() / 1e6;
    auto writeDuration = duration_cast<microseconds>(writeEnd - writeStart).count() / 1e6;
    auto totalDuration = duration_cast<microseconds>(writeEnd - readStart).count() / 1e6;
    
    // Print timings
    cout << "File: " << inputFile << " (" << data.size() << " records)\n";
    cout << fixed << setprecision(6);  // Set precision for floating point output
    cout << "Read time: " << readDuration << " s\n";
    cout << "Sort time: " << sortDuration << " s\n";
    cout << "Write time: " << writeDuration << " s\n";
    cout << "Total time: " << totalDuration << " s\n";
    
    return 0;
}
